src/umpire/ResourceManager.cpp

```diff
@@ -118,52 +145,105 @@ ResourceManager::initialize()
 #endif
 
 #if defined(UMPIRE_ENABLE_DEVICE)
-  m_memory_resources[resource::Device] = registry.makeMemoryResource("DEVICE", getNextId());
+  {
+    const std::string name = resource::type_to_string(resource::Device);
+    m_memory_resources[name] = registry.makeMemoryResource(name, getNextId());
+  }
 #endif
 
 #if defined(UMPIRE_ENABLE_PINNED)
-  m_memory_resources[resource::Pinned] = registry.makeMemoryResource("PINNED", getNextId());
+  {
+    const std::string name = resource::type_to_string(resource::Pinned);
+    m_memory_resources[name] = registry.makeMemoryResource(name, getNextId());
+  }
 #endif
 
 #if defined(UMPIRE_ENABLE_UM)
-  m_memory_resources[resource::Unified] = registry.makeMemoryResource("UM", getNextId());
+  {
+    const std::string name = resource::type_to_string(resource::Unified);
+    m_memory_resources[name] = registry.makeMemoryResource(name, getNextId());
+  }
 #endif
 
 #if defined(UMPIRE_ENABLE_CUDA)
-  m_memory_resources[resource::Constant] = registry.makeMemoryResource("DEVICE_CONST", getNextId());
+  {
+    const std::string name = resource::type_to_string(resource::Constant);
+    m_memory_resources[name] = registry.makeMemoryResource(name, getNextId());
+  }
 #endif
 
   /*
    * Construct default allocators for each resource
    */
-  auto host_allocator = m_memory_resources[resource::Host];
-  m_allocators_by_name["HOST"] = host_allocator;
-  m_allocators_by_id[host_allocator->getId()] = host_allocator;
+  {
+    const std::string name = resource::type_to_string(resource::Host);
+    auto host_allocator = m_memory_resources[name];
+    m_allocators_by_name[name] = host_allocator;
+    m_allocators_by_id[host_allocator->getId()] = host_allocator;
 
-  m_default_allocator = host_allocator;
+    m_default_allocator = host_allocator;
+  }
+
+#if defined(UMPIRE_ENABLE_NUMA_HOST)
+  {
+    const std::string base_name = "NUMA_NODE_";
+    auto host_nodes = numa::get_host_nodes();
+    for (std::size_t n : host_nodes) {
+      resource::MemoryResourceTraits traits;
+      traits.numa_node = n;
+      const std::string name = base_name + std::to_string(n);
+      m_memory_resources[name] = registry.makeMemoryResource(name, getNextId(), traits);
+    }
+  }
+#endif
 
 #if defined(UMPIRE_ENABLE_DEVICE)
-  auto device_allocator = m_memory_resources[resource::Device];
-  m_allocators_by_name["DEVICE"] = device_allocator;
-  m_allocators_by_id[device_allocator->getId()] = device_allocator;
+  {
+    const std::string name = resource::type_to_string(resource::Device);
+    auto device_allocator = m_memory_resources[name];
+    m_allocators_by_name[name] = device_allocator;
+    m_allocators_by_id[device_allocator->getId()] = device_allocator;
+  }
 #endif
 
 #if defined(UMPIRE_ENABLE_PINNED)
```

Comment:
Can pinned memory be created any other way that cudaMallocHost?

Reply:
Yes, using `cudaHostRegister()`. However, there isn't a pure Linux solution without the CUDA driver involved. A quick investigation of cudaMallocHost shows that it uses mmap with MAP_FIXED to keep the virtual to physical mapping constant. This we can do, but it also maps this into the page table of the GPU, which we don't have direct access to.

Reply:
No problem 😄